simpleclient_httpserver/src/main/java/io/prometheus/client/exporter/HTTPServer.java

```diff
@@ -107,34 +110,109 @@ protected static boolean shouldUseCompression(HttpExchange exchange) {
+    }
+
+    protected final HttpServer server;
 
 
     /**
      * Start a HTTP server serving Prometheus metrics from the given registry.
      */
-    public HTTPServer(InetSocketAddress addr, CollectorRegistry registry) throws IOException {
+    public HTTPServer(InetSocketAddress addr, CollectorRegistry registry, boolean daemon) throws IOException {
         server = HttpServer.create();
         server.bind(addr, 3);
         HttpHandler mHandler = new HTTPMetricHandler(registry);
         server.createContext("/", mHandler);
         server.createContext("/metrics", mHandler);
-        server.setExecutor(Executors.newFixedThreadPool(5));
-        server.start();
+        server.setExecutor(Executors.newFixedThreadPool(5, DaemonThreadFactory.defaultThreadFactory(daemon)));
+        start(daemon);
+    }
+
+    /**
+     * Start a HTTP server serving Prometheus metrics from the given registry using non-daemon threads.
+     */
+    public HTTPServer(InetSocketAddress addr, CollectorRegistry registry) throws IOException {
+        this(addr, registry, false);
     }
 
     /**
      * Start a HTTP server serving the default Prometheus registry.
      */
+    public HTTPServer(int port, boolean daemon) throws IOException {
+        this(new InetSocketAddress(port), CollectorRegistry.defaultRegistry, daemon);
+    }
+
+    /**
+     * Start a HTTP server serving the default Prometheus registry using non-daemon threads.
+     */
     public HTTPServer(int port) throws IOException {
-        this(new InetSocketAddress(port), CollectorRegistry.defaultRegistry);
+        this(port, false);
     }
 
     /**
      * Start a HTTP server serving the default Prometheus registry.
      */
+    public HTTPServer(String host, int port, boolean daemon) throws IOException {
+        this(new InetSocketAddress(host, port), CollectorRegistry.defaultRegistry, daemon);
+    }
+
+    /**
+     * Start a HTTP server serving the default Prometheus registry using non-daemon threads.
+     */
     public HTTPServer(String host, int port) throws IOException {
-        this(new InetSocketAddress(host, port), CollectorRegistry.defaultRegistry);
+        this(new InetSocketAddress(host, port), CollectorRegistry.defaultRegistry, false);
+    }
+
+    /**
+     * Start a HTTP server by making sure that its background thread inherit proper daemon flag.
+     */
+    private void start(boolean daemon) {
+        if (daemon == Thread.currentThread().isDaemon()) {
+            server.start();
+        } else {
+            FutureTask<Void> startTask = new FutureTask<Void>(new Runnable() {
+                @Override
+                public void run() {
+                    server.start();
+                }
+            }, null);
+            DaemonThreadFactory.defaultThreadFactory(daemon).newThread(startTask).start();
+            try {
+                startTask.get();
+            } catch (ExecutionException e) {
+                Throwable cause = e.getCause();
```

Comment:
You can simplify by just rethrowing this. It's probably a port already in use.

Reply:
Do you mean rethrowing ExecutionException? In JDK it is declared as 
```Java
public class ExecutionException extends Exception
```
It is a checked exception and therefore to rethrow it I have to change signature of `start` to 
```Java
private void start(boolean daemon) throws ExecutionException {
```
and then I have to change signature of every constructor of HTTPServer and add the same ExecutionException in it. It will break backward compatibility with all existing clients of  HTTPServer class. 
We know, that `startTask.get();` executes server.start(); which can throw RuntimeException and Errors only, and the best place to convert ExecutionException into the cause RuntimeException or Error is in this place rather then let clients of HTTPServer deal with it.

Reply:
You could rethrow it as a RuntimeException

Reply:
That is exactly what I’m doing here, with a small optimization. This is a standard practice in Java world to analyze cause of execution exception and dethrone it. But if you prefer to wrap a wrapper of runtime exception into yet another runtime exception I can do that to pith this fix forward.

Reply:
@brian-brazil changed per review.